Spanner/src/Session/CacheSessionPool.php

```diff
@@ -313,6 +314,11 @@ public function acquire($context = SessionPoolInterface::CONTEXT_READ)
 
         // If we don't have a session, let's wait for one or throw an exception.
         if (!$session) {
+            if ($exception) {
+                throw $exception instanceof \RuntimeException
+                    ? $exception
+                    : new \RuntimeException($exception->getMessage(), $exception->getCode(), $exception);
+            }
```

Comment:
Should this block go after `$this->waitForNextAvailableSession();`? e.g.

```php
    $session = $this->waitForNextAvailableSession()
    if (!$session && $exception) {
        // throw exception
    } 
```

It seems like there could be a situation where existing functioning code breaks because we start throwing an exception, when it previously was successfully waiting for the next available session. But I'm not super familiar with how this code functions, so I could be wrong!

Reply:
We have an exception thrown from `waitForNextAvailableSession` here: https://github.com/googleapis/google-cloud-php/blob/1b8e3bda738917bea55b88b6404ac36de90fe6d2/Spanner/src/Session/CacheSessionPool.php#L776

We need to override that. We need to think a bit more on this I think.

Reply:
Just to clarify, if we don't have a session at this part of the code:

| Before | With current change | shouldWaitForSession | createSessions returns nonnull Exception |
| ----- | ---- | ----- | ----- |
| RuntimeException after maxCyclesToWaitForSession attempts | ~~***RuntimeException***~~ | T | T|
| RuntimeException after maxCyclesToWaitForSession attempts | RuntimeException after maxCyclesToWaitForSession attempts | T | F|
| RuntimeException | RuntimeException | F | T|
| RuntimeException | RuntimeException | F | F|


How about passing the `$exception` returned from `createSessions` into `waitForNextAvailableSession` (may keep optional, defaults to null) and then reformatting the `RuntimeException` accordingly if the exception isn't null with message and code fields?

Reply:
@vishwarajanand What is that last row in your table stands for?

Reply:
it meant.. before this change or after this change we should see a `RuntimeException`, if `shouldWaitForSession` is false and `createSessions` does NOT return a nonnull exception.